src/main/java/org/wildfly/security/http/HttpAuthenticator.java

```diff
@@ -0,0 +1,229 @@
+    private final boolean required;
+    private final boolean ignoreOptionalFailures;
+    private volatile boolean authenticated = false;
+
+    private HttpAuthenticator(final Supplier<List<HttpServerAuthenticationMechanism>> mechanismSupplier, final HttpExchangeSpi httpExchangeSpi,
+            final boolean required, final boolean ignoreOptionalFailures) {
+        this.mechanismSupplier = mechanismSupplier;
+        this.httpExchangeSpi = httpExchangeSpi;
+        this.required = required;
+        this.ignoreOptionalFailures = ignoreOptionalFailures;
+    }
+
+    /**
+     * Perform authentication for the request.
+     *
+     * @return {@code true} if the call should be allowed to continue within the web server, {@code false} if the call should be
+     *         returning to the client.
+     * @throws HttpAuthenticationException
+     */
+    public boolean authenticate() throws HttpAuthenticationException {
+        return new AuthenticationExchange().authenticate();
+    }
+
+    private boolean isAuthenticated() {
+        return authenticated;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private class AuthenticationExchange extends HttpServerExchange {
+
+        private volatile HttpServerAuthenticationMechanism currentMechanism;
+
+        private volatile int responseCode = -1;
+        private volatile boolean responseCodeAllowed = false;
+
+        /**
+         * @param httpExchangeSpi
+         */
+        AuthenticationExchange() {
+            super(httpExchangeSpi);
+        }
+
+        private boolean authenticate() throws HttpAuthenticationException {
+            List<HttpServerAuthenticationMechanism> authenticationMechanisms = mechanismSupplier.get();
+            try {
+                boolean authenticationAttempted = false;
+                for (HttpServerAuthenticationMechanism nextMechanism : authenticationMechanisms) {
+                    currentMechanism = nextMechanism;
+                    authenticationAttempted = authenticationAttempted | nextMechanism.evaluateRequest(this);
+
+                    if (isAuthenticated()) {
+                        return true;
+                    }
+                }
+                currentMechanism = null;
+
+                if (required || (authenticationAttempted && ignoreOptionalFailures == false)) {
+                    responseCodeAllowed = true;
+                    boolean challengeSent = false;
+                    for (HttpServerAuthenticationMechanism nextMechanism : authenticationMechanisms) {
+                        currentMechanism = nextMechanism;
+                        challengeSent = challengeSent | nextMechanism.prepareResponse(this);
+                    }
+                    currentMechanism = null;
+
+                    if (challengeSent == false && (required || (authenticationAttempted && ignoreOptionalFailures == false))) {
+                        httpExchangeSpi.setResponseCode(FORBIDDEN);
```

Comment:
No if any of the mechanisms had actually added a challenge they would have also specified their response code (challengeSent would also be true) - at this point no mechanisms responded with a challenge so there is no point sending a 401 as that really signals that a challenge has been sent so the remaining options is to halt the entire request.